Move policy to new package, add piping code for OOM adjustment

AnanyaKumar opened the pull request:
@vmarmol Will fill in ApplyOomScoreAdjProcesses and update the PR, but if you have any comments on the file structure, let me know! After adding ApplyOomScoreAdjProcesses, the code should set the OOM score of containers correctly.

vmarmol commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1212,32 +1215,43 @@ func (dm *DockerManager) runContainerInPod(pod *api.Pod, container *api.Containe
+		oomScoreAdj = qos.GetPodOomAdjust(&pod.Spec)
+	}
+	fsManager := fs.Manager{
+		Cgroups: &configs.Cgroup{
```
I'm not sure this will work for all cases. We should get the cgroup from Docker somehow (cAdvisor does this today, maybe we can piggyback off of it)

AnanyaKumar replied:
Ok - by the way, are you sure there's no change of a race with cAdvisor? What if cAdvisor does not register the container before we ask cAdvisor for information?

AnanyaKumar replied:
As I was afraid, there is a race condition if we use cAdvisor. If I sleep for a second, I can get information about the container from cAdvisor, but if I don't, cAdvisor tells me that the docker container doesn't exist. I'm sticking with my original solution, but I'm improving the style and taking out the magic "docker" constant.

edit: I meant if the container sleeps, I tried sleeping but that didn't seem to solve the problem :P

vmarmol replied:
Ah you're right about the race. The current code is wrong for almost all setups though (systemd systems and when we put the containers at root which we do today). Let's try a third approach:
- Get the init PID from Docker (the existing code does this)
- Parse `/proc/<init PID>/cgroup` and find the path for `devices`
- Use that path as the container name

`devices` is the only cgroup Docker guarantees to exist and `/proc/<pid>/cgroup` gives us the current container of that cgroup hierarchy.

AnanyaKumar replied:
Cool, I'll implement this. Thanks! I didn't realize you can get the cgroup a process is in from /proc.

AnanyaKumar replied:
Are we assuming for now that there is a single cgroup hierarchy in the system?

vmarmol commented on pkg/kubelet/qos/policy.go:
```diff
@@ -0,0 +1,52 @@
+// (barring exceptional cases). Giving over a 1000 difference between top tier and best effort
+// pods means that best-effort pods will be killed before top-tier pods.
+const (
+	PodInfraOomAdj   int = -501
```
I'd make this even smaller for PodInfraContainers. Else it would be a victim in top-tier as well.

AnanyaKumar replied:
I'm making it -701. I don't want to make it lower, because the kubelet/kube-proxy are more critical, so I need to reserve a lower oom adjust score for those critical tasks. What do you think?

vmarmol replied:
SGTM

vmarmol commented on pkg/kubelet/qos/policy.go:
```diff
@@ -0,0 +1,52 @@
+
+// Policy contains functions that decide what quality of service a pod should get.
+
+// isBestEffort returns true if the pod is a best-effort pod, and false if it is a top-tier pod.
```
nit: IsBestEffort()

AnanyaKumar replied:
Fixed!

vmarmol commented on pkg/kubelet/qos/policy.go:
```diff
@@ -0,0 +1,52 @@
+	return false
+}
+
+// getPodOomAdjust returns the amount by which the OOM score of all processes in the pod should be adjusted.
```
nit: GetPodOomAdjust()

AnanyaKumar replied:
Fixed!

vmarmol commented on pkg/util/oom.go:
```diff
@@ -0,0 +1,60 @@
+// Writes 'value' to /proc/<pid>/oom_score_adj for all processes produced by processLister.
+// Keeps trying to write until processLister produces the same list, or until max retries.
+// The main use case of this is to set oom_score_adj for all processes in some group (e.g. cgroup).
+func ApplyOomScoreAdjProcesses(processLister func() ([]int, error), oomScoreAdj, maxRetries int) error {
```
Maybe take a container name (or a cgroup hierarchy) instead of a lister? Makes for an easier API and more like what the user intention is.

AnanyaKumar replied:
I agree with you, I'll add a function to oom.go that takes a cgroup hierarchy. Initially I implemented it this way because I was worried about a cyclic dependency which turned out not to exist :P

AnanyaKumar commented:
Made requested changes, refactored the code, and implemented the function that sets the OOM score of processes in a container. I'll test the code now.

AnanyaKumar commented:
I ran pods with resources and without resources, and it sets the oom scores of processes in the pods correctly. My plan is to write unit tests later, because right now we aren't sure if the community is happy with our design. Let me know what you think!

vmarmol commented on pkg/util/oom.go:
```diff
@@ -0,0 +1,97 @@
+		pidStr = strconv.Itoa(pid)
+	}
+
+	oom_value, err := ioutil.ReadFile(path.Join("/proc", pidStr, "oom_score_adj"))
```
Consider just setting the OOM score without worrying about the current value. Would simplify the code and not necessarily add overhead.

AnanyaKumar replied:
Done

vmarmol commented on pkg/util/oom.go:
```diff
@@ -0,0 +1,97 @@
+	if err != nil {
+		return fmt.Errorf("failed to read oom_score_adj: %v", err)
+	} else if string(oom_value) != strconv.Itoa(value) {
+		if err := ioutil.WriteFile(path.Join("/proc", pidStr, "oom_score_adj"), []byte(strconv.Itoa(value)), 0700); err != nil {
```
If we keep this, can we not re-create the path :)

AnanyaKumar replied:
Done.

vmarmol commented on pkg/util/oom.go:
```diff
@@ -0,0 +1,97 @@
+		// TODO: look into fork specifications/implementation to see if this is necessary.
+		time.Sleep(processListingDelay)
+	}
+	return fmt.Errorf("Exceeded maxTries, some processes might not have desired OOM score.")
```
nit: s/Exceeded/exceeded/ since it's an error

AnanyaKumar replied:
Done.

vmarmol commented on pkg/util/oom.go:
```diff
@@ -0,0 +1,97 @@
+		}
+		// Sleep, because a process might have forked just before we wrote its OOM score adjust.
+		// The forked process id might not be reflected in cgroup.procs for a short amount of time.
+		// TODO: look into fork specifications/implementation to see if this is necessary.
```
Really? This one I would not expect. I would not expect us to need the sleep.

AnanyaKumar replied:
I see, I'm not familiar with the implementation of fork - I'll talk to you IRL tomorrow :)

AnanyaKumar commented:
I was looking into cgroups, how to create them, and docker's support for them yesterday/today (looks like I'll have to use the --cgroup-parent feature that @vishh added!) - so I'll put up the more sophisticated CPU qos code in the next PR.

vmarmol commented:
Yes, this is a good start though. Behavior will be near-identical. Just some nits, this looks good overall.

vmarmol commented:
LGTM, let's squash and merge :)
